decode/area2png.1
." $Id: area2png.1,v 1.0 2004/09/16 10:00:00 yoksas Exp $
.TH AREA2PNG 1 "2004/09/16 10:00:00 $"
.SH NAME
area2png \- PNG compression routine for image data in McIDAS AREA
file format

.SH SYNOPSIS
.HP
.ft B
area2png
.nh
\%[-a\ \fISATANNOT\fP]
\%[-b\ \fISATBAND\fP]
\%[-f\ \fIAREAfile\fP]
\%[-n\]
\%[-v]
\%[-x]
\%[-l\ \fIlogpath\fP]
\%\fIPNGfile\fP
.hy
.ft

.SH DESCRIPTION
.LP
This program is an LDM encoder for McIDAS data of type \fBAREA\fP:
it reads a McIDAS AREA file and outputs a PNG compressed file that
is composed of the AREA file header(s) followed by AREA file comment
cards followed by compressed image information.
.LP
This program is typically run before injecting imagery into the Unidata
IDD.

.SH FLAGS
.TP
.BI "-a " SATANNOT
McIDAS satellite description file, SATANNOT, pathname (default: /home/mcidas/data/SATANNOT).
.TP
.BI "-b " SATBAND
McIDAS satellite band description file, SATBAND, pathname (default: /home/mcidas/data/SATBAND).
.TP
.BI "-f " AREAfile
Input AREA pathname (required: no default).
.TP
.B "-h "
Print help information, then exit the program.
.TP
.B "-n "
Echo the output file name to stdout (default=no).
.TP
.B -v
Verbose logging.
Informative messages (level LOG_INFO) are logged.
By default, only messages of severity LOG_NOTICE and greater are
logged. See the Unidata  LDM Site  Manager's  Guide  for more information on 
message logging.
.TP
.B -x
Debug logging.
Debugging messages (level LOG_DEBUG) are logged. See the Unidata  LDM Site  
Manager's  Guide  for more information on message logging.
.TP
.BI "-l " logpath
Log file pathname.
The program uses Unidata's \fBulog\fP(3) package to write error and diagnostic
messages.
The default is to write to the \fBsyslogd\fP(8) daemon.  If this option is
specified, however, then messages will be written to the file
\fIlogpath\fP unless that pathname is "-", in which case all messages will be
written to standard error.
.TP
.BI "PNGfile "
Output PNG pathname mask (required: no default).
.sp
PNGfile may contain 'replaceables':
  \\BAND - band information from SATBAND
  \\RES  - resolution information from SATBAND
  \\SAT  - satellite platform information from SATANNOT
.sp
and 'strftime' style day/time specifiers that will get resolved by image
day/time information:
  %%Y    - image year       [CCYY]
  %%m    - image month      [MM]
  %%d    - image day        [DD]
  %%j    - image Julian day [JJJ]
  %%H    - image hour       [HH]
  %%M    - image minute     [MM]
  etc.

.SH EXAMPLES
Typical command invocations might look like:
.sp
.ta 8n +8n +8n
.nf
##
# Example 1:
#
#   System has McIDAS-X Version 7.6 or higher installed in the
#   'mcidas' account under /home/mcidas.
#
#   The file to be comressed is /data/mcidas/AREA1234.
#
#   The user wants to put the compressed output into the file
#   'pngarea' in the current directory.
#
#   Informational notices are logged by the system daemon, syslogd
#   to its standard location for local0.

area2png -f /data/mcidas/AREA1234 pngarea

##
# Example 2:
#
#   User has put the copies of SATANNOT and SATBAND that are
#   distributed with ldm-mcidas into the /home/ldm/etc directory.
#
#   The file to be comressed is /data/mcidas/AREA1234.
#
#   The user wants to put the compressed output into the file
#   'pngarea' in the current directory.
#
#   Informational notices are logged to the users session.
#

area2png -a /home/ldm/etc/SATANNOT -b /home/ldm/etc/SATBAND
         -f /data/mcidas/AREA1234 -vl- pngarea

##
# Example 3:
#
#   System has McIDAS-X Version 7.6 or higher installed in the
#   'mcidas' account under /home/mcidas.
#
#   The file to be comressed is /data/mcidas/AREA1234.
#
#   The user wants to save the compressed data in a file whose
#   name is dictated by by information in the AREA file header
#   and the files SATANNOT and SATBAND.
#
#   Informational and debug notices are logged to the users session.
#

area2png -f /data/mcidas/AREA1234 -xl- \SAT_\BAND_\RES_%Y%m%d.%H%M

The output file name from such an invocation might look like:

GOES-8_IMG_10.7um_4km_200000613.1215

##
# Example 4:
#
#   System has McIDAS-X Version 7.6 or higher installed in the
#   'mcidas' account under /home/mcidas.
#
#   The file to be comressed is /data/mcidas/AREA1234.
#
#   The user wants to save the compressed data with a pathname
#   whose directory components are dictated by information
#   in the AREA file header and the files SATANNOT and SATBAND.
#
#   Informational, debugging, and verbose logging is made to
#   the user's session.
#

area2png -f /data/mcidas/AREA1234 -vxl- \SAT/\RES/\BAND/\BAND-%Y%m%d.%H%M

The output file pathname from such an invocation might look like:

GOES-8_IMG/4km/10.7um/10.7um-200000613.1215
.fi
.RE
.SH "SEE ALSO"
.LP
.BR ldmd (1),
.BR ulog (3),
.BR syslogd (8).


RELEASE_NOTES
VERSION	COMMENTS
-------	-------------------------------------------------------------

2012    2012-01-18

        Updated version number to 2012 to reflect small updates (e.g.,
        new src/port/cfortran/cfortran.h, modified src/decode/Makefile)
        and that the package is still viable.

2008    2009-04-30

        Note: The version number is kept at 2008 since McIDAS went
        into maintenance mode with v2008.

        A variety of changes to C routines, Makefiles and test data
        files.

        Introduction of a new McIDAS lightning decoder, 'lgt2md', that
        handles both NLDN binary and USPLN ASCII data record formats.
        'lgt2md' is designed as a drop-in replacement for 'nldn2md':
        user's pattern action file entries for decoding NLDN data
        can be easily updated by replacing 'nldn2md' with 'lgt2md'.

        ROUTE.SYS (distributed as route.sys) was updated to include
        a new product, 'LP', for USPLN lightning data.

        The Makefile in the src/decode directory was reorganized to
        build the subset of the encoders/decoders contained in the package
        that are written entirely in C:

        area2png    - PNG-compresses the image portion of single-banded
                      McIDAS AREA files
        pnga2area   - uncompresses images created by 'area2png'
        zlibg2gini  - uncompresses the Zlib-compressed images contained
                      in the NOAAPort broadcast (and conveyed in the
                      IDD in the NIMAGE feed)
        pngg2gini   - uncompress NOAAPort images that have been
                      PNG-compressed by the Unidata NOAAPort ingest
                      package (archaic)
        proftomd    - decodes wind profiler netCDF files from NOAA/GSD
                      into McIDAS MD format files
        lgt2md      - decodes NLDN and USPLN lightning data into
                      McIDAS MD files

        This set is built by running 'make' or 'make basic'.
        
        In addition to this set, the Fortran-based 'nldn2md' NLDN
        lightning decoder will be built by 'make full'.

        The Makefile in src/decode also provides for the (attempted)
        building of archaic McIDAS decoders:

        cdftomd     - decodes wind profiler netCDF files from NOAA/GSD
                      into McIDAS MD format files (use 'proftomd' instead)
        gunrv2      - decodes compressed McIDAS GRID files sent in the
                      old Zephyr/Alden satellite broadcast into McIDAS
                      GRID format
        lwfile
        lwtmd2      - decodes compressed McIDAS MD files sent in the
                      old Zephyr/Alden satellite broadcast into McIDAS
                      MD format
        lwtoa3      - decodes delta-compressed McIDAS image files from
                      the old Zephyr/Alden satellite broadcast into McIDAS
                      AREA format
        nids2area   - converts Nexrad Level III images into McIDAS
                      AREA file format (best to use McIDAS IMGCOPY from
                      NEXR to AREA datasets directly as McIDAS capabilities
                      are kept up to date)
 
        In addition decoders built by 'make full', an attepmt will be
        made to build these decoders when the user runs 'make all'.
        There is _no_ guarantee that these decoders will be built
        successfully!

2004    2004-09-14

        Added O_TRUNC flag to mkdirs_open call in pnga2area.c,
        pngg2gini.c, and zlibg2gini.c.  This corrects a problem
        encountered when decoding Unidata-Wisconsin images (IDD
        feedtype UNIWISC aka MCIDAS) when using McIDAS routing table
        (ROUTE.SYS) concepts.  The problem was manifested in decoded
        image files keeping the size of the images they were
        overwriting.  So, the file size was always the "high water
        mark" of sizes for the products.

        NOTES:
        - this did not affect the sizes of files decoded without using
          McIDAS routing table concepts
        - the decoded images using McIDAS routing table concepts were
          valid/usable even though they are larger than they need to
          be

        Changed IDD datastream name from MCIDAS to UNIWISC in example
        pqact.conf files.

        Updated distribution with the versions of SATANNOT, SATBAND,
        and SCHEMA from the latest Unidata McIDAS distribution, v2004.

        Cleaned up batch.k shell script that runs McIDAS postprocess
        BATCH files.  Internal comments should now be easier to follow
        than in previous versions.

        Updated man pages for currently used decoders: area2png(1),
        nldn2md(1), pnga2area(1), pngg2gini(1), proftomd(1), and
        zlibg2gini(1).

        Removed building of decoders now considered to be obsolete:
        cdftomd(1), lwtoa3(1), gunrv2(1), lwfile(1), lwtmd2(1),
        and nids2area(1).  Source code for these decoders is still
        included in the distribution, but their use is not guaranteed
        mainly since subroutines they use from the McIDAS library
        are no longer supported.

2003    2003-07-29

        Added utililty routine zlibg2gini that uncompresses Zlib
        compressed GINI images in NIMAGE IDD datastream.

        Fixed a bug that kept McIDAS routing table, ROUTE.SYS,
        dates and times from being updated.

2002    2002-12-20

        Replaced information messages in pngg2gini.c that referred to
        pnga2area with pngg2gini.

        Corrected logic in pnga2area.c for processing a product whose
        product code is not contained in the McIDAS routing table,
        ROUTE.SYS.  Before, the routing table would get corrupted
        and the AREA file number included in the '-r pcode,anum'
        flag would not be used.  Now, the routing table will not
        be updated for unregistered products, and the default AREA
        file number will be used.
	

decode/area2png.c
#define UD_FORTRAN_SOURCE

#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <time.h>

#include "ulog.h"
#include "png.h"

#define IMG_DIR_LEN 64

char            *annotfile=NULL;
char            *bandfile=NULL;

char             mcpathdir[]="./";                /* McIDAS MCPATH string */

int              READ_ERROR=0;

int              rtecom[64];     /* supply definition for getrte_ and fshrte_ */

png_size_t       pngcount=0;

/*
** Function prototypes
*/

char *GetFileName( int *imgdir, char *format );
void  fbyte4_( void *buffer, int *num );
void  fshrte_(  int *, int * );
int   getrte_(  int *, int *, char * );
int   ischar ( const void *value );

static void
usage( const char *av0 )
{
#define USAGE_FMT "\
Usage: %s -f AREAfile <-a SATANNOT> <-b SATBAND> <-vxl-> PNGfile\n\
Where:\n\
    -a SATANNOT  McIDAS SATANNOT pathname (def: /home/mcidas/data/SATANNOT)\n\
    -b SATBAND   McIDAS SATBAND pathname (def: /home/mcidas/data/SATBAND)\n\
    -f AREAfile  Input AREA pathname (no default)\n\
    -h           Print the help file, then exit the program\n\
    -n           Echo the output file name to stdout (def=no)\n\
    -l logname   Log file name (default uses local0; use '-' for stdout)\n\
    -v           Verbose logging\n\
    -x           Debug logging\n\
Positionals:\n\
    PNGfile      Output PNG pathname mask (no default)\n\n\
                 'PNGfile' may contain 'replaceables':\n\
                    \\BAND - band information from SATBAND\n\
                    \\RES  - resolution information from SATBAND\n\
                    \\SAT  - satellite platform info from SATANNOT\n\n\
                 and 'strftime' style day/time specifiers:\n\
                    %%Y    - image year       [CCYY]\n\
                    %%m    - image month      [MM]\n\
                    %%d    - image day        [DD]\n\
                    %%j    - image Julian day [JJJ]\n\
                    %%H    - image hour       [HH]\n\
                    %%M    - image minute     [MM]\n\
                    etc.\n\n\
Example Invocation:\n\
    ./area2png -l - -f /data/ldm/mcidas/AREA0170 \\SAT_\\RES_\\BAND_%%Y%%m%%d.%%H%%M\n\n\
might expand to the equivalent:\n\
    ./area2png -l - -f /data/ldm/mcidas/AREA0170 GOES-8_IMG_8km_6.8um_20000511.2000\n\n"

    (void) fprintf( stderr, USAGE_FMT, av0 );
}

static char     *logfname = "";

int
main(int argc, char *argv[])
{
    char        *infile=NULL;

    char        *prodmmap;
    char        *memheap;

    char        *pathname;

    char         defannotfile[] = "/home/mcidas/data/SATANNOT";
    char         defbandfile[]  = "/home/mcidas/data/SATBAND";

    int          ch;
    int          echo=0;
    int          fdi;
    int          fdo;
    int          imgdir[64];
    int          i;
    int          one=1;

    int          nelem;
    int          nline;
    int          elesiz;
    int          nband;
    int          presiz;
    int          linlen;
    int          imgsiz;
    int          datoff;
    int          cmtlen;
    int          cmtoff;

    int          nbytes;
    int          nwords;
    int          outlen;

    mode_t       mode = S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH;

    struct stat  statb;

    extern int   optind;
    extern int   opterr;
    extern char *optarg;

    int          logmask = (LOG_MASK(LOG_ERR) | LOG_MASK(LOG_NOTICE)) ;
    int          logfd;

    /*
    ** Validate command line arguments
    */

    if ( argc < 4 ) {
      usage( argv[0] );
      exit( 0 );
    }

    /*
    ** Parse command line flags
    **
    ** NOTE: getopt args expecting a value need to be followed by a colon.
    **
    **       -f /data/mcidas/AREA0120
    */

    opterr = 1;
    while ((ch = getopt(argc,argv,"hna:b:f:vxl:")) != EOF) {
      switch(ch) {
        case 'h':
        case '?':
                 usage( argv[0] );
                 exit( 0 );
                 break;
        case 'a':
                 annotfile = optarg;
                 break;
        case 'b':
                 bandfile = optarg;
                 break;
        case 'f':
                 infile = optarg;
                 break;
        case 'n':
                 echo = 1;
                 break;
        case 'v':
                 logmask |= LOG_MASK(LOG_INFO);
                 break;
        case 'x':
                 logmask |= LOG_MASK(LOG_DEBUG);
                 break;
        case 'l':
                 if ( optarg[0] == '-' && optarg[1] != 0 ) {
                   (void) fprintf( stderr, "logfile \"%s\" ??\n", optarg );
                   usage( argv[0] );
                   exit( 0 );
                 }
                 /* else */
                 logfname = optarg;
                 break;

      }
    }

    /*
    ** Sanity checks
    */

    if ( argc - optind < 0 ) {
      usage( argv[0] );
      exit( 0 );
    }

    if ( infile == (char *) NULL ) {
      printf( "No input file specified" );
      exit( 1 );
    }

    if ( argc != (optind + 1) ) {
      printf( "No output file specified" );
      exit( 1 );
    }

    /*
    ** Fill in default SATANNOT and SATBAND pathnames if needed
    */

    if ( annotfile == (char *) NULL ) annotfile = defannotfile;
    if ( bandfile  == (char *) NULL ) bandfile = defbandfile;

    /*
    ** Setup syslog logging.
    */

    (void) setulogmask( logmask );

    if ( logfname == NULL || !(*logfname == '-' && logfname[1] == 0) )
      (void) fclose( stderr );

    logfd = openulog( ubasename(argv[0]), (LOG_CONS|LOG_PID), LOG_LDM,
                      logfname );

    if ( logfname == NULL || !(*logfname == '-' && logfname[1] == 0) ) {
      setbuf( fdopen(logfd, "a" ), NULL);
    }

    /*
    ** Start
    */

    unotice( "Starting Up" );

    /*
    ** Open the input data file specified on the command line
    */

    fdi = open( infile, O_RDONLY, 0 );

    if ( fdi == -1 ) {
      uerror( "Can't open input file %s", infile );
      exit( 1 );
    }

    if ( fstat(fdi, &statb) == -1 ) {
      uerror( "Can't fstat input file %s", infile );
     (void) close( fdi );
     exit( 2 );
    }

    udebug( "input AREA size: %d", (int)statb.st_size );

    /*
    ** Memory map the input file
    */

    prodmmap = (char *) mmap( 0, statb.st_size,
                              PROT_READ, MAP_PRIVATE, fdi, 0 );

    if ( (void *) prodmmap == (void *) -1 ) {
      perror("prodmmap: allocation failed");
      (void) close(fdi);
      exit( 3 );
    }

    /*
    ** Determine if the input file is a McIDAS AREA
    */

    nwords = IMG_DIR_LEN;
    nbytes = 4 * nwords;

    (void) memcpy( imgdir, prodmmap, nbytes );

    if ( imgdir[0] != 0 ) {
      uerror( "Input file %s fails word 0 AREA test", infile );
      exit ( 4 );
    }

    if ( imgdir[1] != 4 ) {
      fbyte4_( &imgdir[1], &one );

      if ( imgdir[1] != 4 ) {
        uerror( "Input file %s fails word 1 AREA test", infile );
        exit( 5 );
      }

      for ( i=2; i<64; i++ ) {           /* flip bytes in non-character words */
        if ( !ischar( &imgdir[i] ) ) fbyte4_( &imgdir[i], &one );
      }
    }

    /*
    ** Flesh out the output file name and verify that it can be created.
    */

    pathname = GetFileName( imgdir, argv[optind] );
    if ( pathname == (char *) NULL ) {
      uerror( "Error creating output name from pattern: %s", argv[optind] );
      (void) close( fdi );
      exit( 6 );
    }

    fdo = mkdirs_open( pathname, O_WRONLY|O_CREAT|O_TRUNC, mode );
    if ( fdo < 0 ) {
      uerror( "Error opening/creating output file name %s", pathname );
      exit( 1 );
    }

    if ( echo )
      printf( "%s", pathname );

    uinfo( "output file pathname: %s", pathname );

    /*
    ** Allocate memory for the PNG compressed output file.  We assume that
    ** worst compression possible will result in an output file size the
    ** same as the input file size.
    */

    memheap = (char *) malloc( statb.st_size );
    if ( memheap == (char *) NULL ) {
      uerror( "Error mallocing %d bytes for compressed output file",
               (int)statb.st_size );
      (void) close( fdi );
      exit( -7 );
    }

    /*
    ** Begin transferring image data to the compressed output file
    */

    udebug( "memory mapped input file %s", infile );

    png_set_memheap( memheap );

    /*
    ** Get the shape of the image(s) contained in the AREA.
    */

    nline  = imgdir[ 8];                    /* number of image lines        */
    nelem  = imgdir[ 9];                    /* number of image elements     */
    elesiz = imgdir[10];                    /* # bytes per element          */
    nband  = imgdir[13];                    /* # image bands                */
    presiz = imgdir[14];                    /* length of line prefix        */
    
    linlen = presiz + nband * elesiz * nelem;
    imgsiz = nline * linlen;

    udebug( "image shape: %d x %d", nline, nelem );
    udebug( "image line length: %d", linlen );
   
    datoff = imgdir[33];
    cmtlen = 80 * imgdir[63];
    cmtoff = datoff + imgsiz;

    udebug( "number of bytes to beginning of data: %d", datoff );
    udebug( "number of bytes in comment cards: %d", cmtlen );
    udebug( "number of bytes to beginning of comment cards: %d", cmtoff );

    /*
    ** Copy all of the image headers to the output file
    */

    png_header( prodmmap, datoff );         /* img, nav, cal, aux dirs      */
    png_header( prodmmap+cmtoff, cmtlen );  /* comment cards                */

    /*
    ** Initialize PNG output routines, and write out compressed image lines.
    */

    pngout_init( linlen, nline ); 

    for ( i = 0; i < nline; i++)
       pngwrite((unsigned char *)(prodmmap+datoff+(i*linlen)));

    /*
    ** Close off the PNG compression
    */
    pngout_end(); 
    close( fdi );

    outlen = png_get_prodlen();

    udebug( "deflated length %d", outlen );

    unotice( "doPNG:: %8d  %8d  %6.4f\n", (int)statb.st_size,
           outlen, ((float) outlen) / statb.st_size );

    /*
    ** Write the output file
    */

    write( fdo, memheap, outlen );
    close( fdo );

    exit( 0 );
}
